util/collections.py

```diff
@@ -0,0 +1,344 @@
+        if mime_type is not None:
+            criteria.append((MIME_TYPE_KEY, mime_type))
+        if len(criteria) == 0:
+            raise ValueError('At least one of "content" or "mime-type" has to be provided')
+        for index, column in enumerate(self.schema):
+            matched = 0
+            for field, value in criteria:
+                if column[field] == value:
+                    matched += 1
+            if matched == len(criteria):
+                return index
+        return -1
+
+    def read_row(self, row_index, *columns):
+        columns = list(columns)
+        column_data = [None] * len(columns)
+        found = 0
+        if not 0 <= row_index < len(self.offsets):
+            raise ValueError('Wrong sample index: {} - has to be between 0 and {}'
+                             .format(row_index, len(self.offsets) - 1))
+        self.sdb_file.seek(self.offsets[row_index] + INT_SIZE)
+        for index in range(len(self.schema)):
+            chunk_len = self.read_int()
+            if index in columns:
+                column_data[columns.index(index)] = self.sdb_file.read(chunk_len)
+                found += 1
+                if found == len(columns):
+                    return tuple(column_data)
+            else:
+                self.sdb_file.seek(chunk_len, 1)
+        return tuple(column_data)
+
+    def __getitem__(self, i):
+        audio_data, transcript = self.read_row(i, self.speech_index, self.transcript_index)
+        transcript = transcript.decode()
+        sample_id = self.sdb_filename + ':' + str(i)
+        return CollectionSample(sample_id, self.audio_type, audio_data, transcript)
+
+    def __iter__(self):
+        for i in range(len(self.offsets)):
+            yield self[i]
+
+    def __len__(self):
+        return len(self.offsets)
+
+    def close(self):
+        if self.sdb_file is not None:
+            self.sdb_file.close()
+
+    def __del__(self):
+        self.close()
+
+
+class CSV:
+    def __init__(self, csv_filename):
+        self.csv_filename = csv_filename
+        self.rows = []
+        csv_dir = Path(csv_filename).parent
+        with open(csv_filename, 'r') as csv_file:
+            reader = csv.DictReader(csv_file)
+            for row in reader:
+                wav_filename = Path(row[COLUMN_FILENAME])
+                if not wav_filename.is_absolute():
+                    wav_filename = csv_dir / wav_filename
+                self.rows.append((str(wav_filename), int(row[COLUMN_FILESIZE]), row[COLUMN_TRANSCRIPT]))
+        self.rows.sort(key=lambda r: r[1])
+
+    def __getitem__(self, i):
+        wav_filename, _, transcript = self.rows[i]
+        with open(wav_filename, 'rb') as wav_file:
+            return CollectionSample(wav_filename, AUDIO_TYPE_WAV, wav_file.read(), transcript)
+
+    def __iter__(self):
+        for i in range(len(self.rows)):
+            yield self[i]
+
+    def __len__(self):
+        return len(self.rows)
+
+
+class Interleaved:
```

Comment:
Could you elaborate on that class ?

Reply:
It interleaves a bunch of other sample-collections to always return the current shortest sample of all "connected" collections as the next one during iteration.

Reply:
That would definitively deserve a bit of docstring then, I completely failed to get that.